### PR TITLE
Feat(Instagoric): Custom Denoms Support through Faucet

### PR DESCRIPTION
**Pull request by Muneeb147**
### Video Clip:

https://github.com/user-attachments/assets/7a392a93-ec8d-43bb-a6ca-8541ec551972

### Description:
**Background/Problem:** In our faucet app, we send funds to an address and also fund the provision pool. There are only two options `send IST` and `send BLD/IBC` with a hardcoded amount in code. So we can't send custom denoms if needed.

**Solution:** This PR shows a list of denoms (fetched from chain) and show them as a third custom option. Minor refactoring at server is also done (Pre-defined commands types, PROVISIONING_POOL_ADDR is picked from server rather than client, common functions etc..). The clip attached above can further explain the use-case.

### CODE REVIEW

**Comment by Muneeb147** [comment]
Resolved conflicts and tested it with @usmanmani1122 's TX polling change as well. Worked fine

@toliaqat @usmanmani1122 @frazarshad

File: bases/shared/instagoric-server/server.js [hunk 1]
```diff
@@ -11,11 +11,20 @@ import { makeSubscriptionKit } from '@agoric/notifier';
...
+
+const COMMANDS = {
+  "SEND_BLD/IBC": "send_bld_ibc",
+  "SEND_AND_PROVISION_IST": "send_ist_and_provision",
+  "FUND_PROV_POOL": "fund_provision_pool",
+  "CUSTOM_DENOMS_LIST": "custom_denoms_list",
+};
+
+
+const PROVISIONING_POOL_ADDR = 'agoric1megzytg65cyrgzs6fvzxgrcqvwwl7ugpt62346';
```
Comment (toliaqat) [inline bases/shared/instagoric-server/server.js:26]: This address might be different for different testnets? I am surprized that it had worked so far for all of them (emerynet, devnet, other testnets)

Comment (Muneeb147) [reply bases/shared/instagoric-server/server.js:26]: This a provision reserved account address regardless of any testnet. Here's a screenshot of agoric-sdk:

Screenshot 2024-09-12 at 12 43 17 AM

Btw I just moved this address from client (html) to server. Was already hardcoded.

Comment (Muneeb147) [reply bases/shared/instagoric-server/server.js:26]: Link:
https://github.com/Agoric/agoric-sdk/blob/07f6885c560333981d6564581b2a00735a6c428c/packages/internal/src/config.js#L51

File: bases/shared/instagoric-server/server.js [hunk 2]
```diff
@@ -548,30 +583,86 @@ privateapp.listen(privateport, () => {
...
 
-faucetapp.get('/', (req, res) => {
+
+faucetapp.get('/', async (req, res) => {
+
+  const denoms = await getDenoms();
+  let denomHtml = '';
+  denoms.forEach((denom) => {
+    denomHtml += `<label><input type="checkbox" name="denoms" value=${denom}> ${denom} </label>`;
+  })
```
Comment (usmanmani1122) [inline bases/shared/instagoric-server/server.js:593]: Are we sure there are no denoms we want to exclude? cc: @toliaqat

Comment (toliaqat) [reply bases/shared/instagoric-server/server.js:593]: These are testnets and I don't worry about that.

File: bases/shared/instagoric-server/server.js [hunk 3]
```diff
@@ -490,33 +499,59 @@ const sendFunds = async (address, amount) => {
...
+          [exitCode, txHash] = await sendFunds(address, DELEGATE_AMOUNT || constructAmountToSend(BASE_AMOUNT, ['ubld', ...ibcDenoms]));
           break;
         }
+        case COMMANDS["FUND_PROV_POOL"]: {
+          [exitCode, txHash] = await sendFunds(PROVISIONING_POOL_ADDR, DELEGATE_AMOUNT || constructAmountToSend(BASE_AMOUNT, denomsInChain));
+          break;
+        }
+
+        case COMMANDS["CUSTOM_DENOMS_LIST"]: {
+          [exitCode, txHash] = await sendFunds(address, DELEGATE_AMOUNT || constructAmountToSend(BASE_AMOUNT, Array.isArray(denoms) ? denoms : [denoms]));
```
[the remark this answers is not in the document]
Comment (toliaqat) [reply bases/shared/instagoric-server/server.js:551]: We never send any tokens in mainnet. All these changes are for testnets.

File: bases/shared/instagoric-server/server.js [hunk 4]
```diff
@@ -548,30 +583,86 @@ privateapp.listen(privateport, () => {
...
+</select>`
     : '';
   res.send(
-    `<html><head><title>Faucet</title></head><body><h1>welcome to the faucet</h1>
+    `<html><head><title>Faucet</title>
+    <script>
+    function toggleRadio(event) {
+            var field = document.getElementById('denoms');
+
+            if (event.target.value === "custom_denoms_list") {
```
[the remark this answers is not in the document]
Comment (Muneeb147) [reply bases/shared/instagoric-server/server.js]: Done

File: bases/shared/instagoric-server/server.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (usmanmani1122) [reply bases/shared/instagoric-server/server.js:551]: Then custom list shouldn't give `DELEGATE_AMOUNT` priority right?

[the remark this answers is not in the document]
Comment (Muneeb147) [reply bases/shared/instagoric-server/server.js:551]: Well touseef can answer it better that what behaviour we want to keep. Does env always need to be prioritized? It'll look buggy if we select custom denoms but backend use env (but this is current behavior for radio selections of other tokens too).

**Comment by Muneeb147** [comment]
@usmanmani1122 @toliaqat As per our discussion, the generic denoms are reverted back and we're using hardcoded denoms for the two radio buttons (send IST and provision and send BLD/IBC toy tokens). And they are preferring env variable.
BUT for custom denoms list option, env variable (DELEGATE_AMOUNT) is not respected otherwise having custom denoms option is of no use.

Sounds fine? Changes are tested on deployed k8 namespace.

**Review by usmanmani1122 — APPROVED** [review]

**Pull request merged**